lib/puppet/environments.rb

```diff
@@ -213,9 +213,34 @@ def create_environment(name)
     end
 
     def valid_directory?(envdir)
-      name = Puppet::FileSystem.basename_string(envdir)
-      Puppet::FileSystem.directory?(envdir) &&
-         Puppet::Node::Environment.valid_name?(name)
+      return false unless Puppet::Node::Environment.valid_name?(Puppet::FileSystem.basename_string(envdir))
+
+      # Make an attempt to access the directory to determine if it's invalid
+      # due to permission problems or simply because it doesn't exist
+      begin
+        stat = Puppet::FileSystem.stat(envdir)
+        if stat && stat.directory?
+          raise Errno::EACCES, envdir.to_s unless stat.readable?
+          true
+        else
+          false
+        end
+      rescue Errno::ENOENT
+        false
+      rescue Errno::EACCES
+        # Report the directory where the actual access error occurs (traverse parents)
+        prev = envdir
+        current = Puppet::FileSystem.dir(prev)
+        while current
+          begin
+            break if Puppet::FileSystem.stat(current).readable?
+          rescue Errno::EACCES
+          end
+          prev = current
+          current = Puppet::FileSystem.dir(prev)
+        end
+        raise Errno::EACCES, prev.to_s
+      end
```

Comment:
We reverted this commit in https://github.com/puppetlabs/puppet/pull/5667. I think just cherry-pick the revert?